modules/vault/src/main/java/org/apache/ignite/internal/vault/impl/VaultServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.ignite.lang.IgniteUuid;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Simple in-memory representation of vault. Only for test purposes.
+ */
+public class VaultServiceImpl implements VaultService {
+    /** Map to store values. */
+    private TreeMap<ByteArray, byte[]> storage = new TreeMap<>();
+
+    /**
+     * Special key for vault where applied revision for {@code putAll} operation is stored.
+     */
+    private static ByteArray APPLIED_REV = ByteArray.fromString("applied_revision");
+
+    /** Mutex. */
+    private final Object mux = new Object();
+
+    private final WatcherImpl watcher;
+
+    public VaultServiceImpl() {
+        this.watcher = new WatcherImpl();
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<VaultEntry> get(ByteArray key) {
+        synchronized (mux) {
+            return CompletableFuture.completedFuture(new VaultEntry(key, storage.get(key)));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @NotNull @Override public CompletableFuture<Long> appliedRevision() {
+        synchronized (mux) {
+            return CompletableFuture.completedFuture(IgniteUtils.bytesToLong(storage.get(APPLIED_REV), 0));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<Void> put(ByteArray key, byte[] val) {
+        synchronized (mux) {
+            storage.put(key, val);
+
+            watcher.notify(new VaultEntry(key, val));
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<Void> remove(ByteArray key) {
+        synchronized (mux) {
+            storage.remove(key);
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public Iterator<VaultEntry> range(ByteArray fromKey, ByteArray toKey) {
+        synchronized (mux) {
+            return new ArrayList<>(storage.subMap(fromKey, toKey).entrySet())
+                .stream()
+                .map(e -> new VaultEntry(e.getKey(), e.getValue()))
+                .iterator();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<IgniteUuid> watch(@NotNull VaultWatch vaultWatch) {
+        synchronized (mux) {
+            return watcher.register(vaultWatch);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Void> stopWatch(@NotNull IgniteUuid id) {
+        synchronized (mux) {
+            watcher.cancel(id);
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Void> putAll(@NotNull Map<ByteArray, byte[]> vals, long revision) {
+        synchronized (mux) {
+            byte[] appliedRevBytes = storage.get(APPLIED_REV);
+
+            long appliedRevision = appliedRevBytes != null ? IgniteUtils.bytesToLong(appliedRevBytes, 0) : 0L;
+
+            if (revision < appliedRevision)
```

Comment:
Instead of silently ignoring `revision < appliedRevision` I'd rather return false in order to notify user about safe but unexpected behavior.

Reply:
There are no any revision in vault. It is mixing of business logic and storage functionality. Must be removed to the manager.

Reply:
We agreed with @sanpwc to throw `CheckedException` when `revision < appliedRevision`. Also agreed, that this should be moved to manger